File: App/Sources/KnottySession/Mouse.swift
import CKnotty

/// Which way a mouse moved.
public enum MouseAction: Sendable {
    /// A button went down.
    case press
    /// A button came back up.
    case release
    /// The pointer moved.
    case motion

    var raw: UInt8 {
        let action =
            switch self {
            case .press: KT_MOUSE_ACTION_PRESS
            case .release: KT_MOUSE_ACTION_RELEASE
            case .motion: KT_MOUSE_ACTION_MOTION
            }
        return UInt8(action.rawValue)
    }
}

/// Which button a mouse event is about.
public enum MouseButton: Sendable {
    /// No button, which only a motion can be.
    case none
    /// The left button.
    case left
    /// The right button.
    case right
    /// The middle button.
    case middle

    /// Which button `NSEvent.buttonNumber` names, or nil for one no terminal
    /// protocol has a code for.
    ///
    /// AppKit numbers them in the order the protocols do not: 1 is the right
    /// button and 2 the middle. Past those are the side buttons, which a
    /// terminal could carry — the engine numbers eleven — and which nothing
    /// here has yet been asked for.
    public init?(buttonNumber: Int) {
        switch buttonNumber {
        case 0: self = .left
        case 1: self = .right
        case 2: self = .middle
        default: return nil
        }
    }

    var raw: UInt8 {
        let button =
            switch self {
            case .none: KT_MOUSE_BUTTON_NONE
            case .left: KT_MOUSE_BUTTON_LEFT
            case .right: KT_MOUSE_BUTTON_RIGHT
            case .middle: KT_MOUSE_BUTTON_MIDDLE
            }
        return UInt8(button.rawValue)
    }
}

/// How many whole lines a wheel has turned, and the fraction of one it is
/// still holding.
///
/// A trackpad reports its inertia in pixels and reports a great many of them:
/// a single flick is hundreds of events, most of them a fraction of a line.
/// The core is told in lines, so what is left over between two calls has to
/// wait somewhere — and here is where, because the height a line is drawn at
/// is what turns one into the other and that belongs to the display side.
///
/// A wheel with detents reports in lines already. macOS ramps the magnitude of
/// those with how fast the wheel is spun and reports the slowest turn as a
/// tenth of a line, so a turn is rounded up to the one line it plainly was —
/// truncating it instead means a slow wheel does nothing at all.
public struct WheelLines: Sendable {
    /// The fraction of a line each axis is holding, in points.
    private var pending = (x: 0.0, y: 0.0)

    public init() {}

    /// Take a scroll event, and answer the whole lines it completed.
    ///
    /// `(0, 0)` for a precise event that did not finish a line, which is most
    /// of them — and the call that does not happen because of it is the point
    /// of this type.
    public mutating func lines(
        deltaX: Double,
        deltaY: Double,
        precise: Bool,
        cellSize: (width: Double, height: Double)
    ) -> (x: Int, y: Int) {
        guard precise else {
            return (x: detents(deltaX), y: detents(deltaY))
        }
        pending.x += deltaX
        pending.y += deltaY
        return (x: whole(&pending.x, cellSize.width), y: whole(&pending.y, cellSize.height))
    }

    /// The whole lines `pending` has come to, taken out of it.
    private func whole(_ pending: inout Double, _ line: Double) -> Int {
        // A cell is never no pixels, but a view mid-layout can still say so,
        // and dividing by it would answer with infinity. What was held is
        // dropped with it: a remainder that can never drain is one that grows
        // for as long as the view is that size.
        guard line > 0 else {
            pending = 0
            return 0
        }
        guard pending.magnitude >= line else { return 0 }
        let lines = (pending / line).rounded(.towardZero)
        pending -= lines * line
        return Int(lines)
    }

    /// What a wheel with detents turned, which is a count of lines already.
    private func detents(_ delta: Double) -> Int {
        guard delta != 0 else { return 0 }
        let whole = Int(delta.rounded(.towardZero))
        return delta > 0 ? max(1, whole) : min(-1, whole)
    }
}

File: crates/knotty-core/src/mouse.rs
//! What a mouse did before anything has decided what bytes it becomes.
//!
//! The cell rather than the pixel: converting one to the other wants the
//! metrics, and those live where the display is. Everything after that —
//! whether the child hears about it at all, which of the five reporting
//! formats it is written in, whether a wheel is a mouse code or a cursor key
//! or the viewport moving — is read against modes the terminal holds and so
//! is decided beside it. cf. `docs/adr/0017-semantic-input-events.md`

/// Which way a mouse moved.
#[repr(u8)]
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum MouseAction {
    /// A button went down.
    Press = 0,
    /// A button came back up.
    Release = 1,
    /// The pointer moved.
    Motion = 2,
}

/// Which button a mouse event is about.
///
/// The three a terminal has ever been told about, and the absence that a
/// motion with nothing held is. The engine names eight more; nothing on this
/// side has a way to press one, so nothing here names them.
#[repr(u8)]
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum MouseButton {
    /// No button — which only a motion can be.
    None = 0,
    /// The left button.
    Left = 1,
    /// The right button.
    Right = 2,
    /// The middle button.
    Middle = 3,
}

/// One mouse event, as it crosses the boundary: what happened to which
/// button, what was held at the time, and which cell it happened over.
///
/// Undecided on purpose. A click with reporting off comes to no bytes at all,
/// and which of the reporting formats a click that does report is written in
/// is the terminal's to say — reading either anywhere but beside the terminal
/// reads it as of some earlier frame. cf.
/// `docs/adr/0017-semantic-input-events.md`
#[derive(Clone, Copy, Debug)]
pub struct MouseEvent {
    /// Which way the mouse moved.
    pub action: MouseAction,
    /// Which button it was about.
    pub button: MouseButton,
    /// What was held down, as [`crate::key::Modifier`] bits.
    pub mods: u16,
    /// The column the pointer was over, counted from the left of the
    /// viewport.
    pub x: u16,
    /// The row the pointer was over, counted from the top of the viewport.
    pub y: u16,
}

/// One turn of the wheel, in lines, over one cell.
///
/// Lines rather than pixels: a trackpad reports its inertia a few pixels at
/// a time and reports a great many of them, and turning those into lines
/// wants the height a line is drawn at — which is known where the display
/// is. Up and right are positive.
///
/// Undecided like the rest. What a turn comes to is a mouse code, a run of
/// cursor keys or the viewport moving, and which of the three is a question
/// only the terminal can answer. cf. `docs/adr/0017-semantic-input-events.md`
#[derive(Clone, Copy, Debug)]
pub struct WheelEvent {
    /// How many lines sideways, right positive.
    pub delta_x: i32,
    /// How many lines up or down, up positive.
    pub delta_y: i32,
    /// What was held down, as [`crate::key::Modifier`] bits.
    pub mods: u16,
    /// The column the pointer was over, counted from the left of the
    /// viewport.
    pub x: u16,
    /// The row the pointer was over, counted from the top of the viewport.
    pub y: u16,
}
